src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/GraphQLMiddleware.cs

```diff
@@ -0,0 +1,111 @@
+using System;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using GraphQL;
+using GraphQL.Http;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using Newtonsoft.Json;
+using OrchardCore.Apis.GraphQL.Queries;
+using OrchardCore.Apis.GraphQL.Services;
+
+namespace OrchardCore.Apis.GraphQL
+{
+    public class GraphQLMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly GraphQLSettings _settings;
+        private readonly IDocumentExecuter _executer;
+        private readonly IDocumentWriter _writer;
+
+        public GraphQLMiddleware(
+            RequestDelegate next,
+            GraphQLSettings settings,
+            IDocumentExecuter executer,
+            IDocumentWriter writer)
+        {
+            _next = next;
+            _settings = settings;
+            _executer = executer;
+            _writer = writer;
+        }
+
+        public async Task Invoke(HttpContext context, IAuthorizationService authorizationService, ISchemaService schemaService)
+        {
+            if (!IsGraphQLRequest(context))
+            {
+                await _next(context);
+            }
+            else
+            {
+                var authorized = await authorizationService.AuthorizeAsync(
+                    context.User, Permissions.ExecuteGraphQL);
+
+                if (authorized)
+                {
+                    await ExecuteAsync(context, schemaService);
+                }
+                else
+                {
+                    context.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
+                }
+            }
+        }
+
+        private bool IsGraphQLRequest(HttpContext context)
+        {
+            return context.Request.Path.StartsWithSegments(_settings.Path)
+                && String.Equals(context.Request.Method, "POST", StringComparison.OrdinalIgnoreCase);
+        }
+
+        private async Task ExecuteAsync(HttpContext context, ISchemaService schemaService)
+        {
+            var schema = await schemaService.GetSchema();
+
+            string body;
+            using (var streamReader = new StreamReader(context.Request.Body))
+            {
+                body = await streamReader.ReadToEndAsync().ConfigureAwait(true);
+            }
+
+            var request = JsonConvert.DeserializeObject<GraphQLRequest>(body);
+
+            var queryToExecute = request.Query;
+            if (!String.IsNullOrWhiteSpace(request.NamedQuery))
+            {
+                var namedQueries = context.RequestServices.GetServices<INamedQueryProvider>();
```

Comment:
This is very un-optimal as it creating new dictionaries for every single call. I suggest to change `INamedQueryProvider` to have a `Recolve` method that return the query directly. Do not use dictionaries here. The first implementation for return a result wins.